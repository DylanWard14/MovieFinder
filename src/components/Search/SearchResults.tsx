import React from 'react';
import styled from 'styled-components/macro';
import { Link } from 'react-router-dom';

import LoadingSpinner from '../LoadingSpinner/LoadingSpinner';

const Results = styled.ul`
    width: 100%;
    background: black;
    z-index: 100;
    list-style-type: none;
    padding: 0;
    margin: 0.5rem 0;
    position: absolute;
    color: ${props => props.theme.colors.secondary};
`;

const Result = styled(Link)`
    width: 100%;
    display: inline-block;
    box-sizing: border-box;
    padding: 0.5rem 1rem;
    text-decoration: none;
    color: inherit;
    font-size: 1.125rem;

    :hover,
    :focus {
        background: ${props => props.theme.colors.main};
        color: black;
        cursor: pointer;
    }
`;

const NoData = styled.li`
    padding: 0.5rem 1rem;
    font-size: 1.125rem;
`;

const SearchError = styled.p`
    padding: 0.5rem 1rem;
    font-size: 1.125rem;
    color: ${props => props.theme.colors.error};
`;

interface SearchResultsProps {
    searchStore: {
        loading: boolean;
        searchTerm: string;
        data?: {
            title: string;
            id: number;
        }[];
        error?: Error;
    };
    setVisible: (visible: boolean) => void;
}

const SearchResults: React.FunctionComponent<SearchResultsProps> = ({ searchStore, setVisible }) => {
    if (searchStore.loading) {
        return <LoadingSpinner />;
    }
    if (searchStore.error) {
        return <SearchError>An error occurred, please try again later.</SearchError>;
    }
    return (
        <Results id="search_results">
            {searchStore?.data?.length
                ? searchStore.data.map((item, index) => {
                      return (
                          <li tabIndex={0} key={`${item.title}_search_result_${index}`}>
                              <Result to={`/${item.id}`} onClick={() => setVisible(false)}>
                                  {item.title}
                              </Result>
                          </li>
                      );
                  })
                : searchStore.searchTerm && <NoData>No results found...</NoData>}
        </Results>
    );
};

export default SearchResults;
