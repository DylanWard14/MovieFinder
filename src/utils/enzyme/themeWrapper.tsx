import React from 'react';
import { render, shallow, mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';

import myTheme from '../../theme';

export const renderWrapper = (component: React.ReactNode) => {
    return render(<ThemeProvider theme={myTheme}>{component}</ThemeProvider>);
};

export const shallowWrapper = (component: React.ReactNode) => {
    return shallow(<ThemeProvider theme={myTheme}>{component}</ThemeProvider>);
};

export const mountWrapper = (component: React.ReactNode) => {
    return mount(<ThemeProvider theme={myTheme}>{component}</ThemeProvider>);
};
