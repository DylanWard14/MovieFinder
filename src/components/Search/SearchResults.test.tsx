import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';

import { renderWrapper } from '../../utils/enzyme/themeWrapper';
import SearchResults from './SearchResults';

describe('SearchResults', () => {
    it('renders loading correctly', () => {
        const wrapper = renderWrapper(
            <SearchResults searchStore={{ loading: true, searchTerm: 'test' }} setVisible={() => {}} />,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders with error correctly', () => {
        const wrapper = renderWrapper(
            <SearchResults
                searchStore={{ loading: false, searchTerm: 'test', error: new Error('test error') }}
                setVisible={() => {}}
            />,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders with data correctly', () => {
        const wrapper = renderWrapper(
            <Router>
                <SearchResults
                    searchStore={{
                        loading: false,
                        searchTerm: 'test',
                        data: [
                            { title: 'test1', id: 1 },
                            { title: 'test2', id: 2 },
                            { title: 'test3', id: 3 },
                            { title: 'test4', id: 4 },
                            { title: 'test5', id: 5 },
                        ],
                    }}
                    setVisible={() => {}}
                />
            </Router>,
        );
        expect(wrapper).toMatchSnapshot();
    });
});
