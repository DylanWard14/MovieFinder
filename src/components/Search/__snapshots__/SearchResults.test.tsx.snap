// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults renders loading correctly 1`] = `
<div>
  <div
    class="LoadingSpinner__Ring-nzk750-0 behNsB"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;

exports[`SearchResults renders with data correctly 1`] = `
<ul
  class="SearchResults__Results-miypjy-0 seyyS"
  id="search_results"
>
  <li
    tabindex="0"
  >
    <a
      class="SearchResults__Result-miypjy-1 IggoE"
      href="/1"
    >
      test1
    </a>
  </li>
  <li
    tabindex="0"
  >
    <a
      class="SearchResults__Result-miypjy-1 IggoE"
      href="/2"
    >
      test2
    </a>
  </li>
  <li
    tabindex="0"
  >
    <a
      class="SearchResults__Result-miypjy-1 IggoE"
      href="/3"
    >
      test3
    </a>
  </li>
  <li
    tabindex="0"
  >
    <a
      class="SearchResults__Result-miypjy-1 IggoE"
      href="/4"
    >
      test4
    </a>
  </li>
  <li
    tabindex="0"
  >
    <a
      class="SearchResults__Result-miypjy-1 IggoE"
      href="/5"
    >
      test5
    </a>
  </li>
</ul>
`;

exports[`SearchResults renders with error correctly 1`] = `
<p
  class="SearchResults__SearchError-miypjy-3 bkzDEj"
>
  An error occurred, please try again later.
</p>
`;
